File: frontend/src/api/endpoints/atsIntegration.ts
/**
 * ATI — ATS Integration cross-link bridge API client (ATI-001..ATI-005).
 *
 * Backend prefix: /ui/ats/integration
 * Feature gate:   503 {code: "ats_integration_disabled"} when
 *                 ATS_INTEGRATION_ENABLED=false OR CANDIDATES_ENABLED=false.
 *
 * All endpoints require cookie-auth session (CSRF for non-GET).
 */
import { api } from "@/api/client";

// ─── Inline TypeScript types mirroring app/models.py ─────────────────────────

/** Candidate <-> CRM Contact cross-link. */
export interface CandidateContactLink {
  link_id: string;
  link_type: "candidate_contact";
  candidate_id: string;
  contact_id: string;
  owner_sub: string;
  /** True when the CRM contact was successfully created/fetched from the ATS candidate. */
  synced: boolean;
  /** True when a sibling cluster was unavailable and the link persisted in partial state. */
  degraded: boolean;
  /** Comma-joined reason codes, e.g. "candidates_flag_off,contact_sync_failed" */
  degraded_reason: string | null;
  created_at: number;
  updated_at: number;
}

/** Job Order <-> CRM Opportunity cross-link. */
export interface JobOpportunityLink {
  link_id: string;
  link_type: "job_opportunity";
  job_order_id: string;
  opportunity_id: string;
  owner_sub: string;
  /** True when the CRM opportunity was successfully created/fetched from the job order. */
  synced: boolean;
  /** True when a sibling cluster was unavailable and the link persisted in partial state. */
  degraded: boolean;
  /** Comma-joined reason codes, e.g. "job_orders_flag_off,opportunity_sync_failed" */
  degraded_reason: string | null;
  created_at: number;
  updated_at: number;
}

/** Full links list response from GET /ui/ats/integration/links */
export interface IntegrationLinksOut {
  candidate_contact_links: CandidateContactLink[];
  job_opportunity_links: JobOpportunityLink[];
  count: number;
}

/** Request body for POST /ui/ats/integration/candidate-contact-links */
export interface CandidateContactLinkIn {
  candidate_id: string;
  /** Optional: omit to auto-sync a new contact from the candidate. */
  contact_id?: string;
}

/** Request body for POST /ui/ats/integration/job-opportunity-links */
export interface JobOpportunityLinkIn {
  job_order_id: string;
  /** Optional: omit to auto-sync a new opportunity from the job order. */
  opportunity_id?: string;
}

// ─── API wrappers ─────────────────────────────────────────────────────────────

/** List all ATI cross-links for the current user. */
export const listIntegrationLinks = () =>
  api.get<IntegrationLinksOut>("/ui/ats/integration/links");

/** Create or upsert a Candidate <-> Contact link. */
export const createCandidateContactLink = (body: CandidateContactLinkIn) =>
  api.post<CandidateContactLink>("/ui/ats/integration/candidate-contact-links", body);

/** Create or upsert a Job Order <-> Opportunity link. */
export const createJobOpportunityLink = (body: JobOpportunityLinkIn) =>
  api.post<JobOpportunityLink>("/ui/ats/integration/job-opportunity-links", body);

/**
 * Delete a cross-link by type + id.
 * @param linkType "candidate_contact" | "job_opportunity"
 * @param linkId   The deterministic link_id from the link record.
 */
export const deleteIntegrationLink = (linkType: string, linkId: string) =>
  api.del<{ ok: boolean }>(`/ui/ats/integration/links/${linkType}/${encodeURIComponent(linkId)}`);

File: app/routers/ats_integration.py
"""ATI — ATS Integration cross-link router (ATI-owned).

Thin HTTP surface over ``app/services/ats_integration.py``. All endpoints are
cookie/session auth (CSRF for non-GET) and 503 when the ATI master gate is off.
No sibling router is touched.
"""

from __future__ import annotations

from typing import Dict

from fastapi import APIRouter, Depends, HTTPException

from app.models import (
    AtsCandidateContactLinkIn,
    AtsCandidateContactLinkOut,
    AtsIntegrationLinksOut,
    AtsJobOpportunityLinkIn,
    AtsJobOpportunityLinkOut,
)
from app.services import ats_integration as ati
from app.services.sessions import require_ui_session

router = APIRouter(prefix="/ui/ats/integration", tags=["ats_integration"])


def _guard() -> None:
    if not ati.ats_integration_active():
        raise HTTPException(status_code=503, detail={"code": "ats_integration_disabled"})


@router.post("/candidate-contact-links", response_model=AtsCandidateContactLinkOut)
async def create_candidate_contact_link(
    body: AtsCandidateContactLinkIn,
    ctx: Dict[str, str] = Depends(require_ui_session),
) -> AtsCandidateContactLinkOut:
    _guard()
    row = ati.link_candidate_to_contact(
        owner_sub=ctx["user_sub"],
        candidate_id=body.candidate_id,
        contact_id=body.contact_id,
    )
    return AtsCandidateContactLinkOut(**row)


@router.post("/job-opportunity-links", response_model=AtsJobOpportunityLinkOut)
async def create_job_opportunity_link(
    body: AtsJobOpportunityLinkIn,
    ctx: Dict[str, str] = Depends(require_ui_session),
) -> AtsJobOpportunityLinkOut:
    _guard()
    row = ati.link_job_to_opportunity(
        owner_sub=ctx["user_sub"],
        job_order_id=body.job_order_id,
        opportunity_id=body.opportunity_id,
    )
    return AtsJobOpportunityLinkOut(**row)


@router.get("/links", response_model=AtsIntegrationLinksOut)
async def list_integration_links(
    ctx: Dict[str, str] = Depends(require_ui_session),
) -> AtsIntegrationLinksOut:
    _guard()
    return AtsIntegrationLinksOut(**ati.list_links(ctx["user_sub"]))


@router.delete("/links/{link_type}/{link_id}")
async def delete_integration_link(
    link_type: str,
    link_id: str,
    ctx: Dict[str, str] = Depends(require_ui_session),
) -> Dict[str, bool]:
    _guard()
    ok = ati.unlink(ctx["user_sub"], link_type, link_id)
    if not ok:
        raise HTTPException(status_code=400, detail={"code": "invalid_link_type"})
    return {"ok": True}
